Makefile

```diff
@@ -2270,13 +2279,11 @@ builtin/help.sp builtin/help.s builtin/help.o: EXTRA_CPPFLAGS = \
 	'-DGIT_MAN_PATH="$(mandir_relative_SQ)"' \
 	'-DGIT_INFO_PATH="$(infodir_relative_SQ)"'
 
-version.sp version.s version.o: GIT-VERSION-FILE GIT-USER-AGENT
+version.sp version.s version.o: GIT-VERSION-FILE GIT-USER-AGENT GIT-BUILT-FROM-COMMIT
 version.sp version.s version.o: EXTRA_CPPFLAGS = \
 	'-DGIT_VERSION="$(GIT_VERSION)"' \
 	'-DGIT_USER_AGENT=$(GIT_USER_AGENT_CQ_SQ)' \
-	'-DGIT_BUILT_FROM_COMMIT="$(shell \
-		GIT_CEILING_DIRECTORIES="$(CURDIR)/.." \
-		git rev-parse -q --verify HEAD 2>/dev/null)"'
+	'-DGIT_BUILT_FROM_COMMIT="$(GIT_BUILT_FROM_COMMIT)"'
```

Comment:
Since we take pains to write that file, the value should probably be read from the file instead of relying on the variable, right?

```
-DGIT_BUILT_FROM_COMMIT="$(cat GIT-BUILT-FROM-COMMIT)"
```

But if we do that, we need to change patch 4 to write that file instead of specifying the value in the `make` invocation.

But maybe the idea was to make `GIT_BUILT_FROM_COMMIT` overrideable from the command-line? In that case, I would suggest not to write `GIT-BUILT-FROM-COMMIT` at all, but instead use `?=` when defining the variable (which should avoid multiple evaluations, just like the `:=` idea Jeff pointed out).

Reply:
I think we need the file to be written when the OID changes as she in the Makefile to cause Make to later detect that the dependencies are stale and cause version.o to be recompiled, right?

Reply:
Oh, but of course! The file is there only to force recompilation if `GIT_BUILT_FROM_COMMIT` changed...

Reply:
The idea was, as you suggested, to make `GIT_BUILT_FROM_COMMIT` overrideable on the command line. It mimics some of the other overrideable variables (`GIT_USER_AGENT`, `PERL_DEFINES`, etc.) in the Makefile, which also a) write to a file, and b) do not read from that file when using the value. My interpretation was that the file is used identify whether the value _changed_ on subsequent builds (the "new built-from commit"), and that the Makefile variable is what's actually used. 

Also, this reminded me that I missed something (documentation at the top of the file indicating that `GIT_BUILT_FROM_COMMIT` is overrideable via the command line).